url2a908001cd_malictus_klang_tgz-pJ8-malictus_klang_test_PrimitiveTestJ8/src/malictus/klang/chunk/FORMChunk.java

```
@@ -0,0 +1,212 @@
+			//insert name and size
+			KlangUtil.insertIntoFile(new byte[8], this.getParentFile().getFile(), startingplace);
+			raff.seek(startingplace);
+			FourCC fcc = new FourCC(name);
+			fcc.writeValueToFile(raff);
+			Int4ByteUnsignedBE chunksize = new Int4ByteUnsignedBE();
+			chunksize.setValue(bytes.length);
+			chunksize.writeValueToFile(raff);
+			//now update and reparse
+			((EditableFileBase)this.getParentFile()).reparseFile(raff);
+			this.chunkJustChanged(raff);
+		} catch (BadValueException err) {
+			throw new IOException(KlangConstants.ERROR_GENERAL_FILE);
+		}
+	}
+	
+	public void subChunkWillChangeSize(Chunk chunk, long newsize, RandomAccessFile raf) throws IOException {
+		try {
+			long newDsize = newsize - 8;
+			if (newDsize >= 0) {
+				//write size var to chunk itself
+				Primitive subCsize = new Int4ByteUnsignedBE();	
+				subCsize.setValueFromString(Long.toString(newDsize));
+				raf.seek(chunk.getStartPosition() + 4);
+				subCsize.writeValueToFile(raf);
+			}
+			//now write this chunk's size variable
+			Int4ByteUnsignedBE size = new Int4ByteUnsignedBE();
+			long sizevar = chunk.getEndPosition() - chunk.getStartPosition();
+			//sizevar may be negative now
+			sizevar = newsize - sizevar;
+			long cursize = this.getEndPosition() - this.getStartPosition();
+			size.setValue(cursize + sizevar - 8);
+			raf.seek(startPosition + 4);
+			size.writeValueToFile(raf);
+		} catch (BadValueException err) {
+			err.printStackTrace();
+			throw new IOException(KlangConstants.ERROR_GENERAL_FILE);
+		}
+	}
+	
+	public void subChunkHasChanged(Chunk chunk, RandomAccessFile raf) throws IOException {
+		//for FORM chunks, do nothing
+	}
+	
+	public void editSubChunkName(Chunk chunk, String newName, RandomAccessFile raff) throws IOException {
+		try {
+			raff.seek(chunk.getStartPosition());
+			FourCC fcc = new FourCC();
+			fcc.setValueFromString(newName);
+			fcc.writeValueToFile(raff);
+		} catch (BadValueException err) {
```

Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)